src/http2_connection.erl

```diff
@@ -734,3 +864,136 @@ send_ack_timeout(SS) ->
+            {next_state,
+             handshake,
+             send_settings(NewState)
+            };
+        BadPreface ->
+            lager:debug("Bad Preface: ~p", [BadPreface]),
+            go_away(?PROTOCOL_ERROR, State)
+    end.
+
+%% Incoming data is a series of frames. With a passive socket we can just:
+%% 1. read(9)
+%% 2. turn that 9 into an http2 frame header
+%% 3. use that header's length field L
+%% 4. read(L), now we have a frame
+%% 5. do something with it
+%% 6. goto 1
+
+%% Things will be different with an {active, true} socket, and also
+%% different again with an {active, once} socket
+
+%% with {active, true}, we'd have to maintain some kind of input queue
+%% because it will be very likely that Data is not neatly just a frame
+
+%% with {active, once}, we'd probably be in a situation where Data
+%% starts with a frame header. But it's possible that we're here with
+%% a partial frame left over from the last active stream
+
+%% We're going to go with the {active, once} approach, because it
+%% won't block the gen_server on Transport:read(L), but it will wake
+%% up and do something every time Data comes in.
+
+handle_socket_data(<<>>,
+                   StateName,
+                   #connection_state{
+                      socket={Transport,Socket}
+                     }=State) ->
+    active_once({Transport, Socket}),
+    {next_state, StateName, State};
+handle_socket_data(Data,
+                   StateName,
+                   #connection_state{
+                      socket={Transport, Socket},
+                      buffer=Buffer
+                     }=State) ->
+
+    More = case Transport:recv(Socket, 0, 1) of %% fail fast!
+        {ok, Rest} ->
+            Rest;
+        %% It's not really an error, it's what we want
+        {error, timeout} ->
+            <<>>
+    end,
+
+    %% What is buffer?
+    %% empty - nothing, yay
+    %% {frame, frame_header(), binary()} - Frame Header processed, Payload not big enough
+    %% {binary, binary()} - If we're here, it must mean that Bin was too small to even be a header
+    ToParse = case Buffer of
+        empty ->
+            <<Data/binary,More/binary>>;
+        {frame, FHeader, BufferBin} ->
+            {FHeader, <<BufferBin/binary,Data/binary,More/binary>>};
+        {binary, BufferBin} ->
+            <<BufferBin/binary,Data/binary,More/binary>>
+    end,
+    %% Now that the buffer has been merged, it's best to make sure any
+    %% further state references don't have one
+    NewState = State#connection_state{buffer=empty},
+
+    case http2_frame:recv(ToParse) of
+        %% We got a full frame, ship it off to the FSM
+        {ok, Frame, Rem} ->
+            gen_fsm:send_event(self(), {frame, Frame}),
+            handle_socket_data(Rem, StateName, NewState);
+        %% Not enough bytes left to make a header :(
+        {error, not_enough_header, Bin} ->
+            {next_state, StateName, NewState#connection_state{buffer={binary, Bin}}};
```

Comment:
this means that it's a fragment of a header, and since we can't process a frame header without all nine bytes, we wait until more come over the wire. It's an edge case that probably happens rarely, if at all.